src/app/clusters/drlc-server/drlc-server.cpp

```diff
@@ -0,0 +1,554 @@
+    // https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/8541
+
+    auto & id = commandData.loadControlProgramID;
+    if (id.size() > kProgramIDSize)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    Status status = mDelegate.RemoveLoadControlProgram(id);
+
+    if (status == Status::Success)
+    {
+        MatterReportingAttributeChangeCallback(mEndpoint, Id, LoadControlPrograms::Id);
+    }
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+}
+
+template <bool ClearLoadControlEventsRequestSupported>
+void Instance<ClearLoadControlEventsRequestSupported>::HandleAddLoadControlEventRequest(
+    HandlerContext & ctx, const Commands::AddLoadControlEventRequest::DecodableType & commandData)
+{
+    auto transitionIter = commandData.event.transitions.begin();
+
+    // Check spec-defined constraints on the fields of the struct.
+    auto constraintsSatisfied = [this, &commandData, &transitionIter]() -> bool {
+        auto & event = commandData.event;
+        if (event.eventID.size() != kEventIDSize ||
+            (!event.programID.IsNull() && event.programID.Value().size() != kProgramIDSize) ||
+            event.criticality == CriticalityLevelEnum::kUnknownEnumValue)
+        {
+            return false;
+        }
+
+        size_t transitionCount = 0;
+        while (transitionIter.Next())
+        {
+            ++transitionCount;
+
+            auto transition = transitionIter.GetValue();
+            if (transition.duration > kMaxTransitionDuration)
+            {
+                return false;
+            }
+
+            // TODO: Decide on the best way to check against the feature map here
+            // and reject attempts to add controls that we don't claim to support.
+            // We could have the feature map as a template parameter, but that leads
+            // to a bunch of code duplication if multiple DRLC clusters with
+            // different features are in fact used...  We could store it in a
+            // member, but then we end up paying some codesize costs for the cases
+            // that we don't actually support.  Maybe we should have CHIP_CONFIG
+            // defines for the set of features we might support, and then do runtime
+            // checks.
+            size_t providedControls = 0;
+            if (transition.temperatureControl.HasValue())
+            {
+                ++providedControls;
+                auto & control = transition.temperatureControl.Value();
+                if ((control.coolingTempSetpoint.HasValue() && !control.coolingTempSetpoint.Value().IsNull() &&
+                     control.coolingTempSetpoint.Value().Value() < kMinTemperature) ||
+                    (control.heatingTempSetpoint.HasValue() && !control.heatingTempSetpoint.Value().IsNull() &&
+                     control.heatingTempSetpoint.Value().Value() < kMinTemperature))
```

Comment:
Is there meant to be kMaxTemperature as well?

Reply:
No, the temperature type has a bottom bound (absolute 0), but goes up as high as the storage type allows.